### PR TITLE
Add TrilinosWrappers::SolverBelos

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

File: include/deal.II/lac/trilinos_solver.h [hunk 1]
```diff
@@ -718,8 +727,469 @@ namespace TrilinosWrappers
...
+                    ExcInternalError());
+        AssertThrow(n_cols == static_cast<unsigned int>(this->GetNumberVecs()),
+                    ExcInternalError());
+
+        for (unsigned int i = 0; i < n_cols; ++i)
+          (*this->vectors[i]) *= beta;
+
+        for (unsigned int i = 0; i < n_cols; ++i)
+          for (unsigned int j = 0; j < n_rows; ++j)
+            this->vectors[i]->add(alpha * B(j, i), *A.vectors[j]);
```
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/deal.II/lac/trilinos_solver.h:895]: Correctly, I think we can simply create these general abstractions for blessed vectors (i.e., deal.II's own vectors `Vector` and `LinearAlgebra::distributed::Vector`) and the block vectors building on top of these vectors.

**Review by kronbichler — APPROVED** [review]
I like this interfaces, it gives access to newer solvers.

**Comment by peterrum** [comment]
@kronbichler  I have added some documentation and have worked on the solver control part. Belos only understands relative tolerance so that I need to translate the absolute one.

**Review by kronbichler — APPROVED** [review]
Good. It would be great to have an additional opinion on this PR.

**Review by bangerth — COMMENTED** [review]
I've got to run to get my flight, but here are a few comments already. I'll look at the rest on occasion!

In any case, many thanks!

File: include/deal.II/lac/trilinos_solver.h [hunk 2]
```diff
@@ -718,8 +727,677 @@ namespace TrilinosWrappers
...
+       */
+      static MultiVecWrapper<VectorType> &
+      cast(Belos::MultiVec<value_type> &vec_in)
+      {
+        auto vec = dynamic_cast<MultiVecWrapper<VectorType> *>(&vec_in);
+
+        AssertThrow(vec, ExcInternalError());
+
+        return *vec;
+      }
```
[the remark this answers is not in the document]
Comment (peterrum) [reply include/deal.II/lac/trilinos_solver.h]: I have renamed it to `try_to_get_underlying_vector()`.

File: include/deal.II/lac/trilinos_solver.h [hunk 3]
```diff
@@ -718,8 +727,677 @@ namespace TrilinosWrappers
...
+    a.vmult(r, x);
+    r.sadd(-1., 1., b);
+    const auto norm_0 = r.l2_norm();
+
+    if (solver_control.check(0, norm_0) != SolverControl::iterate)
+      return;
+
+    double relative_tolerance_to_be_achieved =
+      solver_control.tolerance() / norm_0;
+    const int max_steps = solver_control.max_steps();
```
[the remark this answers is not in the document]
Comment (peterrum) [reply include/deal.II/lac/trilinos_solver.h]: I have changed it. The reason for using int was that this is what Trilinos expects. I cast it now when giving the value to the parameter struct.

File: include/deal.II/lac/trilinos_solver.h [hunk 4]
```diff
@@ -718,8 +727,677 @@ namespace TrilinosWrappers
...
+      Teuchos::rcp(new Teuchos::ParameterList(*belos_parameters)));
+
+    belos_parameters_copy->set("Convergence Tolerance",
+                               relative_tolerance_to_be_achieved);
+    belos_parameters_copy->set("Maximum Iterations", max_steps);
+
+    Teuchos::RCP<Belos::SolverManager<value_type, MV, OP>> solver =
+      Teuchos::rcp(
+        new Belos::BlockGmresSolMgr<value_type, MV, OP>(problem,
+                                                        belos_parameters_copy));
```
[the remark this answers is not in the document]
Comment (peterrum) [reply include/deal.II/lac/trilinos_solver.h]: True. I have made gmres a parameter.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**